lang/src/main/kotlin/org/partiql/lang/util/NumberExtensions.kt

```diff
@@ -256,3 +257,302 @@ val Number.isPosInf get() = when (this) {
+    val scale = stripped.scale() - stripped.precision() + 1
+    val scaleAdj = if (scale % 2 == 0) scale else scale - 1
+    val mantissa = stripped.scaleByPowerOfTen(scaleAdj)
+
+    val guess = BigDecimal.valueOf(kotlin.math.sqrt(mantissa.toDouble()))
+
+    // Conservative guess of the precision of the result of a floating point calculation.
+    var guessPrecision = 10
+
+    // we need this additional logic in case of overflow
+    val targetPrecision = mathContext.precision
+    val normalizedPrecision = mantissa.precision()
+    var approx = guess
+
+    val zeroPointFive = BigDecimal.ONE.divide(BigDecimal.valueOf(2))
+    do {
+        // plus 2 for precision buffering
+        val operatingPrecision = kotlin.math.max(
+            kotlin.math.max(guessPrecision, targetPrecision + 2),
+            normalizedPrecision
+        )
+        val tempMC = MathContext(operatingPrecision, RoundingMode.HALF_EVEN)
+        approx = zeroPointFive.multiply(approx.add(mantissa.divide(approx, tempMC), tempMC))
+        // the magic number here is 2p + 2, consider precision(x*x) is maxed at precision(x) + precision(x)
+        guessPrecision = 2 * guessPrecision + 2
+    } while (guessPrecision < targetPrecision + 2)
+
+    // scale modification
+    val unModifiedRes = approx.scaleByPowerOfTen(-scaleAdj / 2).round(MATH_CONTEXT)
+
+    return unModifiedRes.roundToDigits(mathContext)
+}
+
+/**
+ * Computes e^x of a given BigDecimal x.
+ */
+fun BigDecimal.exp(mathContext: MathContext = MATH_CONTEXT): BigDecimal {
+    val operationMC = MathContext(
+        if (2 * mathContext.precision < 0) Int.MAX_VALUE else 2 * mathContext.precision,
+        mathContext.roundingMode
+    )
+    return if (this.signum() == 0) {
+        BigDecimal.ONE.roundToDigits(mathContext)
+    } else if (this.signum() == -1) {
+        val reciprocal = this.negate().expHelper(operationMC)
+        BigDecimal.valueOf(1)
+            .divide(
+                reciprocal,
+                operationMC
+            ).roundToDigits(mathContext)
+    } else {
+        this.expHelper(operationMC).roundToDigits(mathContext)
+    }
+}
+
+/**
+ * Computes the exponential value of a BigDecimal.
+ */
+private fun BigDecimal.expHelper(mathContext: MathContext): BigDecimal {
+    // For faster convergence, we break exponent into integer and fraction parts.
+    // e^x = e^(i+f) = (e^(1+f/i)) ^i
+    // 1 + f/i < 2
+    var intPart = this.divideToIntegralValue(BigDecimal.ONE).setScale(0)
+
+    if (intPart.signum() == 0) {
+        return this.expTaylor(mathContext)
+    }
+
+    val fractionPart = this.subtract(intPart)
+    // 1 + f/i
+    val expInner = BigDecimal.ONE
+        .add(
+            fractionPart.divide(
+                intPart, mathContext
+            )
+        )
+
+    // e^(1+f/i)
+    val etoExpInner = expInner.expTaylor(mathContext)
+    // The build in power function can only handle int type, which max out at 999999999
+    val maxInt = BigDecimal.valueOf(999999999L)
```

Comment:
Where is the max defined? a comment, a constant, etc.?

Reply:
This is a java library built in constraint.  see https://docs.oracle.com/javase/8/docs/api/java/math/BigDecimal.html#pow-int-